pynamodb/models.py

```diff
@@ -886,41 +879,24 @@ def _get_schema(cls) -> Dict[str, Any]:
                     'key_type': RANGE,
                     'attribute_name': attr_cls.attr_name
                 })
-        return schema
-
-    @classmethod
-    def _get_indexes(cls):
-        """
-        Returns a list of the secondary indexes
-        """
-        index_data: Dict[str, List[Any]] = {
-            'global_secondary_indexes': [],
-            'local_secondary_indexes': [],
-            'attribute_definitions': []
-        }
-        for name, index in cls._indexes.items():
-            schema = index._get_schema()
-            idx = {
-                'index_name': index.Meta.index_name,
-                'key_schema': schema['key_schema'],
-                'projection': {
-                    PROJECTION_TYPE: index.Meta.projection.projection_type,
-                },
-            }
+        for index in cls._indexes.values():
+            index_schema = index._get_schema()
             if isinstance(index, GlobalSecondaryIndex):
-                if getattr(cls.Meta, 'billing_mode', None) != PAY_PER_REQUEST_BILLING_MODE:
-                    idx['provisioned_throughput'] = {
-                        READ_CAPACITY_UNITS: index.Meta.read_capacity_units,
-                        WRITE_CAPACITY_UNITS: index.Meta.write_capacity_units
-                    }
-            index_data['attribute_definitions'].extend(schema['attribute_definitions'])
-            if index.Meta.projection.non_key_attributes:
-                idx['projection'][NON_KEY_ATTRIBUTES] = index.Meta.projection.non_key_attributes
-            if isinstance(index, GlobalSecondaryIndex):
-                index_data['global_secondary_indexes'].append(idx)
+                schema['global_secondary_indexes'].append(index_schema)
             else:
-                index_data['local_secondary_indexes'].append(idx)
-        return index_data
+                schema['local_secondary_indexes'].append(index_schema)
+        attr_names = {key_schema[ATTR_NAME]
+                      for index_schema in chain(schema['global_secondary_indexes'], schema['local_secondary_indexes'])
+                      for key_schema in index_schema['key_schema']}
+        attr_keys = {attr.get('attribute_name') for attr in schema['attribute_definitions']}
+        for attr_name in attr_names:
+            if attr_name not in attr_keys:
+                attr_cls = cls.get_attributes()[cls._dynamo_to_python_attr(attr_name)]
+                schema['attribute_definitions'].append({
+                    'attribute_name': attr_cls.attr_name,
+                    'attribute_type': attr_cls.attr_type
+                })
```

Comment:
This is now where we iterate through the attributes used in the index and add them to the attribute definitions.